Kafka

raboczi opened the pull request:
Refactored the "test" method in PredictiveMonitor.py to be callable as a function rather than just printing its result to stdout.  Modified the corresponding test.py script to match this.

This is needed because the (as-yet-uncommitted) clustered version of Nirdizati-runtime sends the prediction to a Kafka topic rather than stdout.

Commit:
The PredictiveMonitor.py modules now output their predictions as function results rather than printing them.

Commit:
The PredictiveMonitor.py modules now output their predictions as function results rather than printing them.

Commit:
Merge branch 'kafka' of https://github.com/nirdizati/nirdizati-runtime into kafka

Commit:
Alternate server that uses Kafka topics to communicate with the various predictive methods.  See README-kafka.md for detailed setup instructions.

Commit:
Error in the documentation about starting join-events-to-predictions.py

andreyrozumnyi commented on PredictiveMethods/CaseOutcome/PredictiveMonitor.py:
```diff
@@ -65,4 +65,4 @@ def test(self, dt_test):
             if label == self.pos_label:
                 # output probability of a positive class
-                print(np.round(predictions_proba[0, label_col_idx], decimals=3))
+                return np.round(predictions_proba[0, label_col_idx], decimals=3)
```
you can do both: print to stdout and return I think. The ideal case would be to preserve the possibility to run the project in dev without kafka and in production with it. Otherwise, it might be difficult to work with it in local, unless we make really good dockerized microservices architecture. What do think?

raboczi replied:
Ideally, the transport between the Node.JS server and the PredictiveMethod python scripts would be abstracted such that it would be easy to switch between local and kafka-mediated.  The main difficulty in doing so is the changes to jobWorker.js.  In the master branch all the asynchronous details are tidily encapsulated within modelRunner's async.parallel block, whereas the kafka branch completely decouples incoming event handling in jobWorker-kafka.js from outgoing event+prediction handling in uiWorker.js.

My underlying justification for the extra complexity is that Node.JS threads are cheap and easy to create transiently for each event, so you can have a thread for event event and predictive method.  The context (which log / sequence_nr / event_nr) is implicit.

However Kafka topics seem(?) heavyweight enough that I create a very few of them statically, explicitly tag the JSON with context (log / sequence_nr / event_nr fields), and do a bunch of routing logic to multiplex events over those shared topics.  Only the logs themselves (BPI12 and BPI17) are organized by means of their own distinct topics, since they exist for the entire execution of Nirdizati and can therefore be created statically.

andreyrozumnyi replied:
I agree that components should be decoupled, thus, we can build microservices architecture. However, note that predictions are sent to UI as soon as they are all available. We cannot send them one by one as clients will see inconsistent data (will not be clear whether all results are up-to-date). From the other side, as soon as server receives event, it already can be used to update all relevant clients. Just be sure that those conditions are satisfied.

Commit:
Added a dataflow diagram to the supplementary README for the kafka branch.

Commit:
Changed the format of the dataflow diagram from SVG to PDF, in hope that browsers will display it inline within README-kafka.md

Commit:
Refactored the dataflow so that join-events-to-predictions.py has an instance for each log.

Commit:
Bug fix: localhost was handcoded as the Kafka producer host rather than using the configured version.

Commit:
Tweaked the ordering of command-line arguments in the Kafka README.

Commit:
Requiring only an existing Zookeeper and Kafka broker, the new deploy-kafka.py script can clear and create the Kafka topics and set up the server and processors.  It will terminate the server and processors when interrupted.

Commit:
Usage error messages use sys.argv[0] rather than hardcoded script names.

Commit:
Added the collator, which currently is actually an identity transformation from event prefixes to event prefixes.

Commit:
Removed the redundant web services in jobWorker-kafka, so that all the back-end work of calculating predictions is done by the Kafka dataflow.  Accordingly removed the old jobWorker-kafka and renamed uiWorker to be the new jobWorker-kafka.
Redis no longer required.  (?)
The Kafka variant of the replayer emits individual events, rather than case prefixes.
Updated the README-kafka documentation to describe the much easier startup procedure using the deploy-kafka script.

Commit:
Fixed typo (deploy-kafka.js should be deploy-kafka.py)

andreyrozumnyi committed:
- revise README to add more information
- fix async script execution

verenich commented:
Please do NOT merge it until this is dealt with https://github.com/nirdizati/nirdizati-runtime/pull/31

verenich committed:
Merge pull request #31 from nirdizati/master

merge master's changes into kafka

andreyrozumnyi committed:
bring kafka version up to date with the recent changes in the master branch (e.g. getSystemState() now requires 3 arguments, not 2 as it did when kafka branched out)

verenich commented:
I thoroughly tested code in the _kafka_ branch, it works fine both locally and remotely. Either way, it can be run via kafka using **deploy-kafka.py** or independently via **deploy.sh**.
Please review it and merge into master
